dataset/shape.cpp

```diff
@@ -141,4 +141,232 @@ Dataset resize(const DatasetConstView &d, const Dim dim,
   return result;
 }
 
+namespace {
+
+/// Stack dims for reshaping one dim into multiple dims
+///
+/// Go through the old dims and:
+/// - if the dim does not equal the dim that is being stacked, copy dim/shape
+/// - if the dim equals the dim to be stacked, replace by stack of new dims
+const Dimensions stack_dims(const Dimensions &old_dims, const Dim from_dim,
+                            const Dimensions &to_dims) {
+  Dimensions new_dims;
+  for (const auto dim : old_dims.labels())
+    if (dim != from_dim)
+      new_dims.addInner(dim, old_dims[dim]);
+    else
+      for (const auto lab : to_dims.labels())
+        new_dims.addInner(lab, to_dims[lab]);
+  return new_dims;
+}
+
+/// Unstack/flatten dims for reshaping multiple dims into one
+///
+/// Go through the old dims and:
+/// - if the dim is contained in the list of dims to be flattened, add the new
+///   dim once
+/// - if not, copy the dim/shape
+const Dimensions unstack_dims(const Dimensions &old_dims,
+                              const Dimensions &from_dims, const Dim to_dim) {
+  Dimensions new_dims;
+  for (const auto dim : old_dims.labels())
+    if (from_dims.contains(dim)) {
+      if (!new_dims.contains(to_dim))
+        new_dims.addInner(to_dim, from_dims.volume());
+    } else {
+      new_dims.addInner(dim, old_dims[dim]);
+    }
+  return new_dims;
+}
+
+/// Either broadcast variable to from_dims before a reshape or not:
+///
+/// 1. If all from_dims are contained in the variable's dims, no broadcast
+/// 2. If at least one (but not all) of the from_dims is contained in the
+///    variable's dims, broadcast
+/// 3. If none of the variables's dimensions are contained, no broadcast
+Variable maybe_broadcast(const VariableConstView &var,
+                         const Dimensions &from_dims) {
+  // TODO: do we need a std::move?
+
+  const auto &var_dims = var.dims();
+
+  bool vardim_found_in_fromdims = false;
+  for (const auto dim : var_dims.labels())
+    if (from_dims.contains(dim))
+      vardim_found_in_fromdims = true;
+
+  bool vardims_contains_all_fromdims = true;
+  for (const auto dim : from_dims.labels())
+    if (!var_dims.contains(dim))
+      vardims_contains_all_fromdims = false;
+
+  // if (need_broadcast && !var_dims.contains(from_dims)) {
+  if (vardim_found_in_fromdims && !vardims_contains_all_fromdims) {
+    Dimensions broadcast_dims;
+    for (const auto dim : var_dims.labels())
+      if (!from_dims.contains(dim))
+        broadcast_dims.addInner(dim, var_dims[dim]);
+      else
+        for (const auto lab : from_dims.labels())
+          // Need to check if the variable contains that dim, and use the
+          // variable shape in case we have a bin edge.
+          if (var_dims.contains(lab))
+            broadcast_dims.addInner(lab, var_dims[lab]);
+          else
+            broadcast_dims.addInner(lab, from_dims[lab]);
+    return broadcast(var, broadcast_dims);
+  } else {
+    return Variable(std::move(var));
+  }
```

Comment:
Is this the same as
```cpp
for (const auto lab : from_dims.labels())
  if (var_dims.contains(lab))
    broadcast_dims.addInner(lab, var_dims[lab]);
  else
    broadcast_dims.addInner(lab, from_dims[lab]);
return broadcast(var, broadcast_dims);
```
?

Reply:
Not quite, because if you are flattening `x` and `y` into `a`, and you pass a 1d coord with dim `z`, this will try to brodcast `z` to `[x, y]` when it should do nothing.

But I did try to simplify the code a little.